### PR TITLE
Add DNS relay for Android

### PR DESCRIPTION
**Pull request by madeye**
It maybe also useful for `ssredir` and `sslocal`.

### CODE REVIEW

File: src/context.rs [hunk 1]
```diff
@@ -268,12 +282,39 @@ impl Context {
         }
     }
 
+    /// Add a record to the reverse lookup cache
+    #[cfg(feature = "dns-relay")]
+    pub fn add_to_reverse_lookup_cache(&self, addr: IpAddr, qname: String) {
+        let mut reverse_lookup_cache = self.reverse_lookup_cache.lock();
+        reverse_lookup_cache.insert(addr, qname);
+        ()
```
Comment (zonyitoo) [inline src/context.rs]: unnecessary `()`

**Comment by zonyitoo** [comment]
Why not just use dns parsers from trust-dns-proto? https://docs.rs/trust-dns-proto/0.19.3/trust_dns_proto/op/query/struct.Query.html

**Comment by zonyitoo** [comment]
There was a `ssdns` binary in this project. But .. `sstunnel` can do exactly the same thing?

**Comment by madeye** [comment]
It's actually an ACL based ChinaDNS, for Android integration.

I'm new to Rust and its libraries. It would be very nice if we can reuse current libraries or code.

**Comment by zonyitoo** [comment]
You could just resolve the conficts. I can refactor it for you.

**Comment by Mygod** [comment]
I am late to the party but why does this look so complicated...?

**Comment by madeye** [comment]
@Mygod Yeah, I didn't take advantage of `trust-dns-proto` package here. It should help to remove most of code for DNS protocol itself.

File: src/relay/sys/unix/bsd_pf.rs [hunk 2]
```diff
@@ -117,7 +124,10 @@ mod ffi {
         }
     }
 
+<<<<<<< HEAD
+=======
     // pub const PF_IN: libc::c_int = 1;
+>>>>>>> upstream/master
```
Comment (Mygod) [inline src/relay/sys/unix/bsd_pf.rs]: It seems like there are more conflicts...

**Comment by Mygod** [comment]
Also that 1000-line rust file is pretty unreadable... (I guess you copied some code from updns?)

I guess I will check back and try to learn rust by example. :)

**Comment by madeye** [comment]
@Mygod Yeah, copy and paste from here: https://github.com/EmilHernvall/dnsguide

**Comment by zonyitoo** [comment]
But, why not put all of theses into a separated binary `ssdns`? It seems that it is not only useful for Android, but for all platforms?

**Comment by madeye** [comment]
@zonyitoo Yeah, I think we can have a `ssdns` for other platforms.

But for android, we prefer only one process, because we need the reverse lookup cache from dns-replay for domain name based ACL rules.

**Comment by zonyitoo** [comment]
Still has confilcts. :(

**Comment by zonyitoo** [comment]
> But for android, we prefer only one process, because we need the reverse lookup cache from dns-replay for domain name based ACL rules.

Then it is a Android only feature. I think we could just simply control it by `#[cfg(target_os = "android")]` instead of a separated feature.

**Comment by madeye** [comment]
I think the `dns-relay` feature is also useful for `ssredir` (or `sstun` one day).

Actually, domain name based ACL only works if we take control of the system's DNS logic.

**Comment by madeye** [comment]
The conflicts should be resolved now.

**Comment by zonyitoo** [comment]
> > But for android, we prefer only one process, because we need the reverse lookup cache from dns-replay for domain name based ACL rules.
> 
> Then it is a Android only feature. I think we could just simply control it by `#[cfg(target_os = "android")]` instead of a separated feature.

How about this? @madeye

File: src/bin/local.rs [hunk 3]
```diff
@@ -34,17 +37,37 @@ fn main() {
...
+                .long("remote-dns")
+                .takes_value(true)
+                .default_value("8.8.8.8:53")
+                .help("Specify the address of remote DNS server (only for Android)"),
         )
         .arg(
-            Arg::with_name("PROTECT_PATH")
-                .long("protect-path")
+            Arg::with_name("DNS_RELAY_ADDR")
+                .long("dns-realy")
```
Comment (Mygod) [inline src/bin/local.rs]: Typo here?

File: src/config.rs [hunk 4]
```diff
@@ -871,13 +871,24 @@ pub struct Config {
     /// ACL configuration
     pub acl: Option<AccessControl>,
     /// Path to stat callback unix address, only for Android
```
Comment (Mygod) [inline src/config.rs]: Forgot to move comment?

File: src/bin/local.rs [hunk 5]
```diff
@@ -183,13 +206,33 @@ fn main() {
...
+        }
+
+        if let Some(dns_relay_addr) = matches.value_of("DNS_RELAY_ADDR") {
+            let addr: SocketAddr = dns_relay_addr.parse().expect("dns relay address");
+            config.dns_relay_addr = Some(addr);
+        }
+    }
+
     if cfg!(target_os = "android") {
+        config.local_dns_path = Some("local_dns_path".to_string());
```
Comment (Mygod) [inline src/bin/local.rs]: This seems unused. Maybe use this on `cfg!(target_os = "android")`? I don't see the necessity of exposing another port.

File: src/relay/dnsrelay/mod.rs [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+async fn udp_lookup(qname: &Name, qtype: RecordType, server: SocketAddr) -> Result<Message> {
+    let mut socket = UdpSocket::bind(("0.0.0.0", 0)).await?;
+
+    let mut message = Message::new();
+    let mut query = Query::new();
+
+    query.set_query_type(qtype);
+    query.set_name(qname.clone());
+
+    message.set_id(6666);
```
Comment (Mygod) [inline src/relay/dnsrelay/mod.rs]: Message ID should be random, no? Same for below.

File: src/relay/dnsrelay/mod.rs [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        }
+    }
+
+    let mut message = Message::new();
+    let mut query = Query::new();
+
+    query.set_query_type(qtype);
+    query.set_name(qname.clone());
+
+    message.set_id(6666);
```
Comment (Mygod) [inline src/relay/dnsrelay/mod.rs]: Also here.

File: src/relay/dnsrelay/mod.rs [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+async fn socks5_lookup(qname: &Name, qtype: RecordType, socks5: SocketAddr, ns: SocketAddr) -> Result<Message> {
+    let mut stream = TcpStream::connect(socks5).await?;
+
+    // 1. Handshake
+    let hs = HandshakeRequest::new(vec![SOCKS5_AUTH_METHOD_NONE]);
+    hs.write_to(&mut stream).await?;
+    stream.flush().await?;
+
+    let hsp = HandshakeResponse::read_from(&mut stream).await?;
+    assert_eq!(hsp.chosen_method, SOCKS5_AUTH_METHOD_NONE);
```
Comment (Mygod) [inline src/relay/dnsrelay/mod.rs]: I think it is okay to send the entire packet at once, which is what `LocalDnsServer` does in shadowsocks-android.

File: src/relay/dnsrelay/mod.rs [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    query.set_query_type(qtype);
+    query.set_name(qname.clone());
+
+    message.set_id(6666);
+    message.set_recursion_desired(true);
+    message.add_query(query);
+
+    let req_buffer = message.to_vec()?;
+    let size = req_buffer.len();
+    let mut size_buffer: [u8; 2] = [((size >> 8) & 0xFF) as u8, ((size >> 0) & 0xFF) as u8];
```
Comment (Mygod) [inline src/relay/dnsrelay/mod.rs]: https://stackoverflow.com/q/25060583/2245107

File: src/relay/dnsrelay/mod.rs [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+    );
+
+    let timeout = Some(Duration::new(5, 0));
+
+    let local_response = try_timeout(udp_lookup(qname, qtype.clone(), local), timeout)
+        .await
+        .unwrap_or(Message::new());
+    let remote_response = try_timeout(socks5_lookup(qname, qtype.clone(), socks5, remote), timeout)
+        .await
+        .unwrap_or(Message::new());
```
Comment (Mygod) [inline src/relay/dnsrelay/mod.rs]: We should send requests first, then await `local_response` and if we know we are going to bypass, we can cancel `remote_response`.

Comment (Mygod) [reply src/relay/dnsrelay/mod.rs]: Of course, this is lower priority...

File: src/relay/dnsrelay/mod.rs (hunk 8 above)
Comment (Mygod) [reply src/relay/dnsrelay/mod.rs]: Also lower priority...

File: src/context.rs [hunk 11]
```diff
@@ -268,12 +282,38 @@ impl Context {
...
+                                if a.check_target_bypassed(self, &reverse_addr).await {
+                                    return true;
+                                }
+                            }
+                        }
+                        _ => (),
+                    }
+                }
+                a.check_target_bypassed(self, target).await
+            }
```
Comment (Mygod) [inline src/context.rs]: `check_target_bypassed` should return `Option<bool>`, and return `None` if nothing matches for hostname.

File: src/relay/dnsrelay/mod.rs [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+
+    let addr = Address::DomainNameAddress(qname.to_string(), 0);
+    let qname_bypassed = context.check_target_bypassed(&addr).await;
```
Comment (Mygod) [inline src/relay/dnsrelay/mod.rs]: `qname_bypassed` should account for `Option<bool>` as mentioned elsewhere...

**Comment by Mygod** [comment]
Re @zonyitoo: in general it is helpful to run a socksifier VPN if we run DNS resolver in the same process as sslocal. If you do not plan to support other platforms for now, I don't see an issue with protecting everything with `cfg`.

**Comment by Mygod** [comment]
Also please keep thing in a pull request so it is easier for me to review them and learn rust. :)

**Comment by zonyitoo** [comment]
> If you do not plan to support other platforms for now

Yes, I have no plan for that.

I have made some modification for cleanup in master branch.

File: src/bin/local.rs (hunk 5 above)
Comment (zonyitoo) [reply src/bin/local.rs]: Yes, this is not used anywhere in code. What is this for? @madeye

**Comment by Mygod** [comment]
That is what I am saying. Please keep the new features in a pull request, this could help me a lot but I guess it is already too late...

We are supposed to use `local_dns_path` instead of `local_dns_addr` on Android so I guess it is still WIP.

**Comment by zonyitoo** [comment]
> I don't see an issue with protecting everything with `cfg`

It is ok to protect it with `cfg`. What I am trying to say is that: we don't need a sepecial feature `dns-relay` to protect that, `target_os = "android"` is already good enough.

**Comment by madeye** [comment]
Okay, I will make this a Android only feature.

File: src/config.rs (hunk 4 above)
Comment (madeye) [reply src/config.rs]: yeah, fixed.

File: src/bin/local.rs (hunk 5 above)
Comment (madeye) [reply src/bin/local.rs]: it's for Mygod's UNIX domain socket based local resolver.

I'll apply `cfg!(target_os = "android")` here.

BTW, I think local socket with UDP (in kernel, it's also implemented like UNIX socket) is already good enough, maybe UNIX socket is not necessary?

File: src/relay/dnsrelay/mod.rs (hunk 6 above)
Comment (madeye) [reply src/relay/dnsrelay/mod.rs]: Fixed.

**Pull request merged**